### PR TITLE
tsuba: bind metadata to RDG

### PR DESCRIPTION
**Pull request by tylershunt**
~`RDG` and `RDGHandle` should be separated, for now here's a hack that uses `Detach` to make a loaded RDG look like an in memory RDG so that it can be written out under a different name.~

Metadata are bound to `RDG` objects rather than Handles.

There are two tensions:
 1. avoiding rewrite of files when we don't have to
 2. making sure current written out property graphs are self contained in directories for portability of test data

I struggled with this for a while because (1) screams for absolute paths, If I write out the graph somewhere else with a different name I can keep all the pointers intact. But I think (2) is to important to ditch.

So in this iteration if you change the "directory" (quotes since directory doesn't really exist on S3 as a concept), everything will be
written our whether it was changed or not.

### CODE REVIEW

File: libtsuba/src/RDG.cpp [hunk 1]
```diff
@@ -45,18 +46,27 @@ struct RDGHandle {
 
   galois::Result<void> Validate() const {
     if (topology_path.empty() || metadata_dir.empty()) {
+      GALOIS_LOG_DEBUG(
```
Comment (witchel) [inline libtsuba/src/RDG.cpp:49]: Will the topology path eventually be an S3 URL?  If so, this logic will have to be modified.

Also, can we say no '/' found, which is more specific then "doesn't look like a path."  Technically speaking a path can absolutely not have a '/'.

File: libtsuba/src/RDG.cpp [hunk 2]
```diff
@@ -540,6 +550,30 @@ RDG::RDG() : handle(std::make_shared<tsuba::RDGHandle>()) {
   edge_table = arrow::Table::Make(arrow::schema({}), empty, 0);
 }
 
+/// dissociate RDG from the handle that was used to create it
```
Comment (witchel) [inline libtsuba/src/RDG.cpp:553]: Detach is just such a weird verb for me.  I Open a graph file.  Then I Load it.  Then I Detach it, Rename it and Store it?  One of these things is different from the others, one of these things doesn't belong.

And on another level, it isn't clear if an RDG has an RDGHandle, or only one, or you have to move it from an RDG to an RDGHandle, or what.

File: libtsuba/src/RDG.cpp [hunk 3]
```diff
@@ -540,6 +550,30 @@ RDG::RDG() : handle(std::make_shared<tsuba::RDGHandle>()) {
...
+galois::Result<void> RDG::Detach() {
+  auto new_handle = std::make_shared<tsuba::RDGHandle>();
+
+  new_handle->node_properties = std::move(handle->node_properties);
+  for (auto& prop : new_handle->node_properties) {
+    prop.path = "";
+  }
+  new_handle->edge_properties = std::move(handle->edge_properties);
+  for (auto& prop : new_handle->edge_properties) {
+    prop.path = "";
```
Comment (witchel) [inline libtsuba/src/RDG.cpp:563]: So if I Detach, then Store, I will get an error saying my paths are empty?  But if I Detach, Rename, Store, then I'm golden?

File: libtsuba/tests/tsuba_dist_write.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  auto rdg = std::move(rdg_res.value());
+  if (auto res = rdg.Detach(); !res) {
+    GALOIS_LOG_FATAL("Detach rdg: {}", res.error());
+  }
+  auto dir_res = galois::CreateUniqueDirectory("/tmp/thunt-");
+  GALOIS_LOG_ASSERT(dir_res);
+  auto dir = dir_res.value();
+
+  std::string meta_file{dir};
+  meta_file += "/meta";
```
Comment (witchel) [inline libtsuba/tests/tsuba_dist_write.cpp:32]: The name of the graph ends in meta?  Why not name the graph after the "directory" that contains it?

File: libtsuba/src/RDG.cpp (hunk 1 above)
Comment (tylershunt) [reply libtsuba/src/RDG.cpp:49]: totally will fix those error messages.

File: libtsuba/src/RDG.cpp (hunk 2 above)
Comment (tylershunt) [reply libtsuba/src/RDG.cpp:553]: Yeah that's the hack part. Right now there's a whole bunch of state bleeding across the abstraction boundary so RDGs and RDGHandles aren't separable like they should be this is what I'm doing to make progress.

File: libtsuba/src/RDG.cpp (hunk 3 above)
Comment (tylershunt) [reply libtsuba/src/RDG.cpp:563]: Yeah, see my above comment

**Comment by tylershunt** [comment]
`IsUri` was also renamed to `IsS3URI`

**Pull request merged**